Zero downtime upgrade docs for plugin writers (#3448)
pulp/pulpcore: docs/plugin_dev/plugin-writer/concepts/index.rst

internationalized will remain in English. This expectation was formed after feedback from
multi-language speakers who believe having error messages for admins in English would reduce the
time to finding a fix and was generally less surprising.

.. _zero-downtime-upgrades:

Zero-Downtime Upgrades
----------------------

Eventually, Pulp users will be able to upgrade without first stopping Pulp services. This has been
`requested from the community <https://discourse.pulpproject.org/t/support-zero-downtime-updates/645>`_,

> To work towards that goal, developers of ``pulpcore`` and ``pulp_file`` should take on the following

Reviewer: I am not sure whether it's worth specifically accentuating on the pulp-file as a plugin, I think every plugin will need to adhere to this requirements.

Author: I revised this a bit more. With "should" remaining in the sentence, having it stated generally as a requirement for all plugins is in-line with the original thinking. +1

Reviewer: yes, thank you.

internationalized will remain in English. This expectation was formed after feedback from
multi-language speakers who believe having error messages for admins in English would reduce the
time to finding a fix and was generally less surprising.

.. _zero-downtime-upgrades:

Zero-Downtime Upgrades
----------------------

As `requested from the community <https://discourse.pulpproject.org/t/support-zero-downtime-updates/645>`_,
in the future, Pulp will likely start allowing upgrades without stopping Pulp services. This will

> cause all Pulp code to take on two new requirements:

Reviewer 2: ```suggestion
cause all Pulp code to take on new requirements:
```
I'm almost sure this list wlll expand...

> * Migrations cannot break earlier versions of running code.

Reviewer 2: ```suggestion
* Migrations must not break earlier versions code possibly still running during an upgrade.
```

Author: I removed the `possibly` since we'll be able to know for sure if its' running or not from the db.

> * Task code must be backwards compatible until Pulp 4.0

Reviewer 2: ```suggestion
* Task code must be backwards compatible until the next major version of Pulp.
```

Future user upgrades will likely run as follow:

> 1. Run the migrations while old pulp code is online. Old code, new data format

Reviewer 2: ```suggestion
1. Run the migrations while old pulp code is online. Old code, is using the new data format.
```

2. Rolling restart old code to become new code. Old and new code is running at the same time!

Zero-Downtime Migrations
========================

The significant challenge with online migrations is that the db state the migration applies has to
work with both newer and older versions of Pulp code. For example, consider a model field that is to

> be renamed. After renaming the model, Django would generate a migration that renames the column to
> be made. This would break all previous code which expects the previous column name.

Reviewer 2: ```suggestion
be renamed. After renaming the field, Django would generate a migration that renames the specifiied
column. This would break all previous code which expects the previous column name.
```

Before getting into specific suggestions, the general pattern is to do the following:

1. The migration should be split into two parts: a "compatible with earlier code migration" and a
   "breaking earlier code migration". In continuing the column rename example, it would become a
   "create a new column" migration, and later a "delete the original column" migration.

2. The "breaking earlier code migration" should be delivered in a later release. It contains the
   the component versions that would not be broken by that change. When that migration goes to run
   it uses the db info in the db for each running pulp process to determine if any components are
   running a version that would break if this change is applied.

The solution to this will be highly dependant on the details of the migration, but here are some
likely patterns to be applied:

> 1. Avoid it. Is this rename really that important? Is it worth the effort?

Reviewer 2: Really just a suggestion...
```suggestion
1. Avoid it. Is this rename really that important? Is it worth the trouble?
```

Author: +1 it's more than just effort it's also uncertainty and support over time. Change made.

2. Rename the model attributes in code and serializers, but leave the actual column name as-is with

>    the `db_column <https://docs.djangoproject.com/en/3.2/ref/models/fields/#db-column>`_.

Reviewer 2: This is almost impossible due to semver on the api anyway, right? But it could be some other api requesting it from us.

Author: Agreed. Based on that, I removed the `and serializers` words since there could be attributes on models we don't expose or expose directly. In those cases I think this idea makes sense. If you think we should modify this more, please lmk.

3. Have an "old" and a "new" column and use database triggers to keep data written to one to also be
   written to the other and vice-versa.

> 4. For data migrations that would normally use table-locks, break these up into chunks.

Reviewer 2: I do not see this as an individual solution to any problem. When you do chunked updates of data, the old code would only see the not yet migrated items, while the new code would only see the already migrated chunks.
It does not sound right to me.

Here's an example:

pulp_file wants to rename a model attribute ``old`` to be called ``new`` in the next pulp_file
release, let's say that's pulp_file 1.10.0. Let's assume that avoiding the rename altogether or
using the ``db-column`` option to just rename the ORM name are not viable.

This could be done as follows:

* Replace the ``old`` field with the ``new`` field, but instead of Django generating a migration
  create one manually that adds the ``new`` column and marks for Django that ``old`` is to be

>   treated as deleted without deleting it.

Reviewer 2: I was wondering if we would keep the column around in the ORM layer and later delete it from there.
I believed this would be much easier with the Django generated migrations.

Author: I agree, I rewrote this part so let me know what you think. Also I realized I never told developers to actually copy the data from `old` to `new` so I added that step in here too.

* The same migration needs to transactionally install a new trigger that anytime ``old`` is written
  to, ``new`` is also written to and vice-versa. For example, something `like this <https://www.appsloveworld.com/postgresql/100/94/how-to-dual-write-to-two-columns-in-one-table-using-postgres-trigger>`_

>   This allows the new code to read/write exclusively with `bar` and the old code to deal with `foo`.

Reviewer 2: ```suggestion
  This allows the new code to read/write exclusively with `new` and the old code to deal with `old`.
```

* Have the codebase of pulp_file 1.10.0 drop the usage of ``old`` entirely.

At a later time, e.g. pulp_file 1.13.0, a migration will be shipped to remove column ``old``. That
migration needs to do two things things:

* Prior to running ensure via the database records that there are no pulp components running with
  pulp_file < 1.10.0. If there are abort, not running the migration and notify the user they need to
  upgrade to a version pulp_file>=1.10,<1.13.0.

> * Remove the column ``old`` and the database trigger.

Reviewer 2: Probably the other way around...

Tasking System
==============

Tasking also has some considerations to allow code upgrades; specifically, tasks dispatched from
older codebases could be runnable on newer, upgraded workers. To ensure this always works tasks must
be forever backwards compatible until Pulp 4.0. For example, you cannot have a breaking signature
change in tasking code and if this is needed you need to make a new task name and preserve the old

> code until Pulp 4.0.

Reviewer 2: I think at some point it should be ok to cancel a very old task with an appropriate error message.

Author: I modified to be `the next major version of Pulp.`. Yes we may need to at some point, but I'd prefer to not add that into the toolbox as an option right from the start if that's ok. Let me know if you still want to.

internationalized will remain in English. This expectation was formed after feedback from
multi-language speakers who believe having error messages for admins in English would reduce the
time to finding a fix and was generally less surprising.

.. _zero-downtime-upgrades:

Zero-Downtime Upgrades
----------------------

As `requested from the community <https://discourse.pulpproject.org/t/support-zero-downtime-updates/645>`_,
in the future, Pulp will likely start allowing upgrades without stopping Pulp services. This will

> cause all Pulp code to take on new requirements:

Reviewer: I am having troubles to understand the goal of these 2 sentences. Is this a future promise but we are still not there? "will likely start allowing" does not give much of the info to the users. How can we re-phrase this better?

Author: I rewrote this paragraph to more clearly state what is hapening. What do you think?

   the `db_column <https://docs.djangoproject.com/en/3.2/ref/models/fields/#db-column>`_.

3. Have an "old" and a "new" column and use database triggers to keep data written to one to also be
   written to the other and vice-versa.

Here's an example:

pulp_file wants to rename a model attribute ``old`` to be called ``new`` in the next pulp_file
release, let's say that's pulp_file 1.10.0. Let's assume that avoiding the rename altogether or
using the ``db-column`` option to just rename it in the ORM are not viable.

This could be done as follows:

* Add the ``new`` field next to the ``old`` field and have Django auto-create a migration adding
  ``new``.

* The same migration needs to install a new trigger that anytime ``old`` is written to, ``new`` is
  also written to and vice-versa. For example, something `like this <https://www.appsloveworld.com/postgresql/100/94/how-to-dual-write-to-two-columns-in-one-table-using-postgres-trigger>`_
  This allows the new code to read/write exclusively with `new` and the old code to deal with `old`.

* Write a data migration that copies ``old`` to ``new`` in bathes. Use batching to avoid a long

>   table-lock.

Reviewer 2: With the trigger in place it may suffice to "update" the old columns.

* Have the codebase of pulp_file 1.10.0 stop using ``old`` entirely.

At a later time, e.g. pulp_file 1.13.0, a migration will be shipped to remove column ``old``. That
migration needs to do two things things:

* Prior to running ensure via the database records that there are no pulp components running with

>   pulp_file < 1.10.0. If there are abort, not running the migration and notify the user they need to
>   upgrade to a version pulp_file>=1.10,<1.13.0.

Reviewer 2: The sentence structure is a little unclear to me. How about:
```suggestion
  pulp_file < 1.10.0. If there are, abort running the migration and notify the user they need to
  upgrade to a version pulp_file>=1.10,<1.13.0.
```

* Remove the database trigger and then the column ``old``.

Data Migrations
===============

One problem that can arise from data migrations is the use of table-locks which would prevent other
code still running from executing concurrently. The typical solution is to have data migrations
operate in transactional batches which avoids a table-lock and instead creates row-locks.

Tasking System
==============

Tasking also has some considerations to allow code upgrades; specifically, tasks dispatched from

> older codebases could be runnable on newer, upgraded workers. To ensure this always works tasks must

Reviewer: `could run` otherwise `they should be runnable` as "-able" points to ability not to the possibility.

> be forever backwards compatible until Pulp 4.0. For example, you cannot have a breaking signature

Reviewer 2: ... the next major version ...

I always try to avoid explicit versions in docs. You will most definitely miss updating them properly in the future.
